### PR TITLE
Kubernetes error indicator

### LINKED ISSUE #13794: Bad Cluster URL Fails To Load All Kubernetes Clusters Resulting in Almost Blank Page
## Expected Behavior
When clusters are configured in backstage, but the URL is incorrect or for a cluster that has been destroyed so the url is pointing to an incorrect address, I would expect those clusters to not appear on the page but other clusters to appear just fine.

## Actual Behavior
If someone deletes a cluster in aws, something that the maintainers of their backstage instance have no control over, it just starts throwing 500s and all kubernetes tab pages fail to load any clusters. The error it says at the top and in the logs is `Error: getaddrinfo ENOTFOUND badurl.gr7.us-west-2.eks.amazonaws.com`. I realize our use case is unique, but we have many clusters owned by many different teams. It is rare, but happens that a cluster in a non-production account may need to be replaced. After that the old url no longer works. All the other cluster's configuration is correct and I would wish that the bad cluster be simply skipped. I would want all other clusters's resources to be queried successfully and displayed correctly. It would be nice to correctly handle this error gracefully. 

Screenshot:
Screen Shot 2022-09-21 at 3 09 19 PM

## Steps to Reproduce
Screenshot above shows an example with a bad url cluster. Steps to somewhat reproduce below, will do my best to include configuration as well.

1. Install [k8s](https://backstage.io/docs/features/kubernetes/overview) backend and frontend plugin following [installation instructions](https://backstage.io/docs/features/kubernetes/installation).
1. [Configure](https://backstage.io/docs/features/kubernetes/configuration) clusters with a similar setup to the one below. Below is the kubernetes part of the app-config.yaml:
    ```yaml
    kubernetes:
      serviceLocatorMethod:
        type: 'multiTenant'
      clusterLocatorMethods:
        - type: 'config'
          clusters:
            - name: clustername
              url: https://badurl.amazonaws.com
              authProvider: 'aws'
              caData: ${CLUSTER_K8S_CONFIG_CA_DATA}
    ```
1. Configure a component to have the appropriate annotations for the k8s plugin. Example catalog-info.yaml file:
    ```yaml
    apiVersion: backstage.io/v1alpha1
    kind: Component
    metadata:
      name: backstage
      description: The Backstage
      annotations:
        backstage.io/kubernetes-label-selector: 'app=backstage'
        backstage.io/kubernetes-namespace: backstage
    spec:
      type: website
      owner: owner
      lifecycle: experimental
    ```

## Context
This is a badly handled error. In the screenshot above only one cluster has a bad url while the others should work fine. Let me know if more information is needed.

## Your Environment
- Browser Information: Google Chrome Version 105.0.5195.102 (Official Build) (arm64)
- Output of `yarn backstage-cli info`: 

```bash
yarn backstage-cli info
yarn run v1.22.17
$ /Users/asaferlich/dev/src/github.com/segmentio/backstage-segment/node_modules/.bin/backstage-cli info
OS:   Darwin 21.6.0 - darwin/arm64
node: v16.14.2
yarn: 1.22.17
cli:  0.18.1 (installed)
backstage:  1.5.1

Dependencies:
  @backstage/app-defaults                          1.0.5
  @backstage/backend-common                        0.15.0
  @backstage/backend-plugin-api                    0.1.1
  @backstage/backend-tasks                         0.3.4
  @backstage/catalog-client                        1.0.4
  @backstage/catalog-model                         1.1.0
  @backstage/cli-common                            0.1.9
  @backstage/cli                                   0.18.1
  @backstage/config-loader                         1.1.3
  @backstage/config                                1.0.1
  @backstage/core-app-api                          1.0.5
  @backstage/core-components                       0.11.0
  @backstage/core-plugin-api                       1.0.5
  @backstage/dev-utils                             1.0.5
  @backstage/errors                                1.1.0
  @backstage/integration-react                     1.1.3
  @backstage/integration                           1.3.0
  @backstage/plugin-analytics-module-ga            0.1.19
  @backstage/plugin-api-docs                       0.8.8
  @backstage/plugin-app-backend                    0.3.35
  @backstage/plugin-auth-backend                   0.15.1
  @backstage/plugin-auth-node                      0.2.4
  @backstage/plugin-catalog-backend-module-github  0.1.6
  @backstage/plugin-catalog-backend                1.3.1
  @backstage/plugin-catalog-common                 1.0.5
  @backstage/plugin-catalog-graph                  0.2.20
  @backstage/plugin-catalog-node                   1.0.1
  @backstage/plugin-catalog-react                  1.1.3
  @backstage/plugin-catalog                        1.5.0
  @backstage/plugin-github-actions                 0.5.8
  @backstage/plugin-github-pull-requests-board     0.1.2
  @backstage/plugin-home                           0.4.24
  @backstage/plugin-kubernetes-backend             0.7.1
  @backstage/plugin-kubernetes-common              0.4.1
  @backstage/plugin-kubernetes                     0.7.1
  @backstage/plugin-org                            0.5.8
  @backstage/plugin-pagerduty                      0.5.1
  @backstage/plugin-permission-common              0.6.3
  @backstage/plugin-permission-node                0.6.4
  @backstage/plugin-permission-react               0.4.4
  @backstage/plugin-proxy-backend                  0.2.29
  @backstage/plugin-scaffolder-backend             1.5.1
  @backstage/plugin-scaffolder-common              1.1.2
  @backstage/plugin-scaffolder                     1.5.0
  @backstage/plugin-search-backend-module-pg       0.3.6
  @backstage/plugin-search-backend-node            1.0.1
  @backstage/plugin-search-backend                 1.0.1
  @backstage/plugin-search-common                  1.0.0
  @backstage/plugin-search-react                   1.0.1
  @backstage/plugin-search                         1.0.1
  @backstage/plugin-shortcuts                      0.3.0
  @backstage/plugin-stack-overflow                 0.1.4
  @backstage/plugin-techdocs-backend               1.2.1
  @backstage/plugin-techdocs-module-addons-contrib 1.0.3
  @backstage/plugin-techdocs-node                  1.3.0
  @backstage/plugin-techdocs-react                 1.0.3
  @backstage/plugin-techdocs                       1.3.1
  @backstage/plugin-user-settings                  0.4.7
  @backstage/release-manifests                     0.0.5
  @backstage/test-utils                            1.1.3
  @backstage/theme                                 0.2.16
  @backstage/types                                 1.0.0
  @backstage/version-bridge                        1.0.1
✨  Done in 0.96s.
```

Note: This issue should probably have the `k8s-plugin` label but I don't think I can add it myself.

### PR DESCRIPTION
**Pull request by jamieklassen**
## Hey, I just made a Pull Request!

Closes #13794.
There's a decent bit of refactoring in here and a lot of the decisions are discussed at some length in the commit messages -- mostly I really wanted to avoid catching any old promise rejection and surfacing it in the UI -- unfortunately I couldn't find a way to narrow to purely `error` events coming from the socket used by `@kubernetes/client-node` to talk to k8s, but I added some logic to make sure that it's at least only promise rejections coming from that library.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/kubernetes/src/components/ErrorPanel/ErrorPanel.tsx [hunk 1]
```diff
@@ -26,10 +30,30 @@ const clustersWithErrorsToErrorMessage = (
...
+                kubernetesFetchError as KubernetesAddrNotFoundError;
+              message = `Could not resolve Kubernetes API hostname: '${hostname}'`;
+              break;
+            }
+            default:
+              message =
+                'message' in kubernetesFetchError
+                  ? `Error communicating with Kubernetes: ${kubernetesFetchError.errorType}, message: ${kubernetesFetchError.message}`
+                  : `Error fetching Kubernetes resource: '${kubernetesFetchError.resourcePath}', error: ${kubernetesFetchError.errorType}, status code: ${kubernetesFetchError.statusCode}`;
+          }
```
Comment (freben) [inline plugins/kubernetes/src/components/ErrorPanel/ErrorPanel.tsx:53]: i'd suggest breaking out all of this into a helper function (maybe even in this same file), so the code below ends up as `{kubernetesErrorToMessage(kubernetesFetchError)}`. Then the function can use `return` on each case instead ofd assigning to a temp variable

File: plugins/kubernetes-common/src/types.ts [hunk 2]
```diff
@@ -223,18 +223,46 @@ export interface PodStatusFetchResponse {
...
+}
+
 /** @public */
 export type KubernetesErrorTypes =
   | 'BAD_REQUEST'
   | 'UNAUTHORIZED_ERROR'
   | 'NOT_FOUND'
   | 'SYSTEM_ERROR'
+  | 'TIMED_OUT'
+  | 'ADDR_NOT_FOUND'
```
Comment (freben) [inline plugins/kubernetes-common/src/types.ts:265]: Isn't it odd to have these two here? Seems to me that the discriminated union above would have more than one case matching it: e.g. if `errorType` is `'TIMED_OUT'`, then it can be EITHER a `KubernetesHttpError` or a `KubernetesTimeoutError` and you have to additionally sub-filter on other properties. I thought that the new error types were kept separate specifically because you wanted to discriminate on them separated from the old "raw" error type

File: .changeset/lemon-chicken-look.md [hunk 3]
```diff
@@ -0,0 +1,10 @@
+---
+'@backstage/plugin-kubernetes': patch
+'@backstage/plugin-kubernetes-backend': patch
+'@backstage/plugin-kubernetes-common': patch
```
Comment (freben) [inline .changeset/lemon-chicken-look.md:4]: seems like breaking changes for the last two here, should be `minor` then

File: plugins/kubernetes-backend/api-report.md [hunk 4]
```diff
@@ -322,6 +324,14 @@ export type KubernetesObjectTypes =
   | 'statefulsets'
   | 'daemonsets';
 
+// @alpha
+export interface KubernetesRejectionHandler {
+  onRejected: (
+    reason: any,
```
Comment (freben) [inline plugins/kubernetes-backend/api-report.md:330]: Should this really be using `any`? We try to avoid it as much as possible.

File: plugins/kubernetes-backend/src/service/KubernetesErrorIndicator.ts [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+      KubernetesErrorTypes
+    > = knownStatusCodeErrors,
+  ) {
+    this.logger = logger;
+    this.statusCodeErrorTypes = statusCodeErrorTypes;
+  }
+  public onRejected: (
+    reason: any,
+    resourcePath?: string,
+  ) => Promise<KubernetesFetchError> = (reason, resourcePath) => {
```
Comment (freben) [inline plugins/kubernetes-backend/src/service/KubernetesErrorIndicator.ts:49]: this shouldn't be an arrow function, just make it a proper member like the other methods

File: plugins/kubernetes-backend/src/service/KubernetesErrorIndicator.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    if (reason.name === 'HttpError' && reason.statusCode) {
+      const { body } = reason as HttpError;
+      this.logger.warn(
+        `statusCode=${
```
Comment (freben) [inline plugins/kubernetes-backend/src/service/KubernetesErrorIndicator.ts:53]: May want to formulate this more as a sentence

File: plugins/kubernetes-backend/src/service/KubernetesErrorIndicator.ts [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+          reason.statusCode
+        } for resource ${resourcePath} body=[${JSON.stringify(body)}]`,
+      );
+      return Promise.resolve({
+        errorType:
+          this.statusCodeErrorTypes[reason.statusCode] || 'UNKNOWN_ERROR',
+        statusCode: reason.statusCode,
+      });
+    }
+    if (this.didIntercept(reason)) {
```
Comment (freben) [inline plugins/kubernetes-backend/src/service/KubernetesErrorIndicator.ts:63]: I am not sure I understand this. Are you ensuring that there's a rejection the first time it's seen but a resolution the second time on? This seems like it's a bit convoluted and potentially hard to follow. Maybe I'm misunderstanding :thinking:

**Review by freben — COMMENTED** [review]
Alright, some initial thoughts. @mclarke47 might want to take a look too for the overall idea

File: plugins/kubernetes-backend/api-report.md (hunk 4 above)
Comment (jamieklassen) [reply plugins/kubernetes-backend/api-report.md:330]: This function gets used as the `onrejected` parameter in `Promise.catch` in a few places, which takes a parameter of type `any` [in the spec](https://github.com/microsoft/TypeScript/blob/896b9228710941f875294ebee2da998d500c8614/src/lib/es5.d.ts#L1515-L1520) -- it's too bad promises don't have an error type like `Exception` in Haskell.

Through super careful reasoning about `@kubernetes/client-node` it could become _sort of_ possible to put a type on it, but there's always a possibility that a crazy exception gets thrown and I'd lean towards making that clear in the type signature (and maybe an explanatory comment).

If you're _really_ against it I'll stare at all the code paths and try to figure out a sensible type, it just might take a while.

**Comment by freben** [comment]
Note the CI output:

```
*************************************************************************************
* You have uncommitted changes to the public API of a package.                      *
* To solve this, run `yarn build:api-reports` and commit all api-report.md changes. *
*************************************************************************************
```

File: plugins/kubernetes-backend/api-report.md (hunk 4 above)
Comment (freben) [reply plugins/kubernetes-backend/api-report.md:330]: Alright. Let's not dwell too much on it then. Does `unknown` work then?

**Comment by freben** [comment]
Ping on this one! I see there are unaddressed comments still.

**Comment by freben** [comment]
@jamieklassen just checking in on this one!

**Comment by jamieklassen** [comment]
@freben I've got a whole new approach that's almost ready over [here](https://github.com/jamieklassen/backstage/commit/e344a55f669d782c01f531d075c51ced354d1cc7) -- I'll probably close this one and open a new one

**Comment by jamieklassen** [comment]
closing in favor of #15250

File: plugins/kubernetes-backend/src/service/KubernetesErrorIndicator.ts (hunk 6 above)
Comment (jamieklassen) [reply plugins/kubernetes-backend/src/service/KubernetesErrorIndicator.ts:53]: I made it a full sentence over here: https://github.com/backstage/backstage/pull/15250/commits/7341131af5c33b1d11378ebb5baa05071caf2eb5

File: plugins/kubernetes-backend/src/service/KubernetesErrorIndicator.ts (hunk 7 above)
Comment (jamieklassen) [reply plugins/kubernetes-backend/src/service/KubernetesErrorIndicator.ts:63]: There's no more indicator and throw-and-catch stuff in the new PR!

File: plugins/kubernetes-backend/api-report.md (hunk 4 above)
Comment (jamieklassen) [reply plugins/kubernetes-backend/api-report.md:330]: There's no more throw-and-catch stuff in the new PR!

File: .changeset/lemon-chicken-look.md (hunk 3 above)
Comment (jamieklassen) [reply .changeset/lemon-chicken-look.md:4]: I've done these bumps in the new PR

File: plugins/kubernetes-common/src/types.ts (hunk 2 above)
Comment (jamieklassen) [reply plugins/kubernetes-common/src/types.ts:266]: Curious to see what you think of this take: https://github.com/backstage/backstage/pull/15250/commits/0ad476a720f520b267e446f2d64c7cafdd8b5462 -- it's definitely a breaking change in the wire format and hopefully has the data modeling properties you wanted. Sadly I wasn't able to distinguish the different error cases as clearly (i.e. can't distinguish a timeout from an ENOTFOUND) but I think it's worth the tradeoff.
